### PR TITLE
redis: complete redis.utils stubs

### PR DESCRIPTION
**Pull request by chagui**
I would like to contribute to #3544 by completing the existing stubs and adding new ones to the `utils` module from the `redis` package.
`from_url` delegates to [Redis.from_url](https://github.com/andymccurdy/redis-py/blob/master/redis/utils.py#L19) so that's what I based the type annotations on.
For `str_if_bytes` and `safe_str` I am not sure if this is the best way to do things so any suggestion/feedback would be nice.

Refrerences:
* redis.utils source: https://github.com/andymccurdy/redis-py/blob/master/redis/utils.py

### CODE REVIEW

File: stubs/redis/redis/utils.pyi [hunk 1]
```diff
@@ -1,8 +1,21 @@
-from typing import Any
+from typing import Any, ContextManager, Optional, Text, overload
+from typing_extensions import Literal
 
-HIREDIS_AVAILABLE: Any
+from .client import Pipeline, Redis
 
-def from_url(url, db=..., **kwargs): ...
-def pipeline(redis_obj): ...
+HIREDIS_AVAILABLE: Any
```
Comment (srittau) [inline stubs/redis/redis/utils.pyi:6]:
```suggestion
HIREDIS_AVAILABLE: bool
```

File: stubs/redis/redis/utils.pyi [hunk 2]
```diff
@@ -1,8 +1,21 @@
...
+@overload
+def from_url(url: Text, db: Optional[int] = ..., *, decode_responses: Literal[True], **kwargs) -> Redis[str]: ...
+@overload
+def from_url(url: Text, db: Optional[int] = ..., *, decode_responses: Literal[False] = ..., **kwargs) -> Redis[bytes]: ...
```
Comment (srittau) [inline stubs/redis/redis/utils.pyi:10]: Use `Any` for `kwargs`.

File: stubs/redis/redis/utils.pyi [hunk 3]
```diff
@@ -1,8 +1,21 @@
...
+@overload
+def str_if_bytes(value: bytes) -> str: ...
+@overload
+def str_if_bytes(value: object) -> object: ...
```
Comment (srittau) [inline stubs/redis/redis/utils.pyi:14]: We should use a type var for the second overload:

```python
_T = TypeVar("_T")
...
@overload
def str_if_bytes(value: _T) -> _T: ...

File: stubs/redis/redis/utils.pyi [hunk 4]
```diff
@@ -1,8 +1,21 @@
...
+@overload
+def safe_str(value: bytes) -> str: ...
+@overload
+def safe_str(value: object) -> str: ...
```
Comment (srittau) [inline stubs/redis/redis/utils.pyi:18]: The first overload is redundant, as `bytes` is a sub-type of `object` and the function returns the same type:

```suggestion
def safe_str(value: object) -> str: ...
```

**Review by srittau — CHANGES_REQUESTED** [review]
Thanks, a few remarks.

File: stubs/redis/redis/utils.pyi (hunk 1 above)
Comment (chagui) [reply stubs/redis/redis/utils.pyi:6]: Good addition!

File: stubs/redis/redis/utils.pyi (hunk 2 above)
Comment (chagui) [reply stubs/redis/redis/utils.pyi:10]: will do

File: stubs/redis/redis/utils.pyi (hunk 4 above)
Comment (chagui) [reply stubs/redis/redis/utils.pyi:18]: Great comment, I was too focused on the underlying use of `str_if_bytes` but the overload is not needed...

File: stubs/redis/redis/utils.pyi (hunk 3 above)
Comment (chagui) [reply stubs/redis/redis/utils.pyi:14]: That was my first thought as well but then I got the following error
> Overloaded function signatures 1 and 2 overlap with incompatible return types

From what I understood, correct me if I am wrong, the generic type could be `bytes` so we end-up with two signatures giving different return type when the argument type is `bytes`. Curious what your thoughts are?

Comment (srittau) [reply stubs/redis/redis/utils.pyi:14]: mypy is a bit overzealous in cases like this and we can disable the warning using `# type: ignore`. Overloads are processed in order, so if `value` is `bytes`, the first overload is matched. (Weirdly, this behavior is not described either in PEP 484 nor the typing docs, but that looks like an oversight. I will bring it up on typing-sig.)

Comment (chagui) [reply stubs/redis/redis/utils.pyi:14]: Interesting, learned something new today! It's a pity it's not documented indeed but I'll update the stubs to use a generic type then.

**Review by srittau — APPROVED** [review]

**Pull request merged**